src/core/properties/positionproperty.cpp

```diff
@@ -28,115 +28,222 @@
-    : PositionProperty(identifier, displayName, Document{}, std::move(position), camera,
+                                   const vec3& position, CoordinateSpace coordinateSpace,
+                                   CameraProperty* camera, InvalidationLevel invalidationLevel,
+                                   PropertySemantics semantics)
+    : PositionProperty(identifier, displayName, Document{}, position, coordinateSpace, camera,
                        invalidationLevel, semantics) {}
 
 PositionProperty::PositionProperty(const PositionProperty& rhs)
-    : CompositeProperty(rhs)
-    , referenceFrame_(rhs.referenceFrame_)
-    , position_(rhs.position_)
-    , positionWorldSpace_(rhs.positionWorldSpace_)
-    , camera_(rhs.camera_) {
-    // Add properties
-    addProperty(referenceFrame_);
-    addProperty(position_);
-
-    referenceFrame_.onChange([this]() { referenceFrameChanged(); });
-    position_.onChange([this]() { positionChanged(); });
+    : CompositeProperty{rhs}
+    , camera_{rhs.camera_}
+    , previousReferenceSpace_{rhs.previousReferenceSpace_}
+    , referenceSpace_{rhs.referenceSpace_}
+    , transformPosition_{rhs.transformPosition_}
+    , position_{rhs.position_}
+
+    , output_{rhs.output_}
+    , worldPos_{rhs.worldPos_}
+    , viewPos_{rhs.viewPos_}
+    , clipPos_{rhs.clipPos_}
+    , screenPos_{rhs.screenPos_} {
+
+    output_.addProperties(worldPos_, viewPos_, clipPos_, screenPos_);
+    addProperties(referenceSpace_, position_, output_);
+
+    util::for_each_argument([&](auto& p) { outputProperties_.push_back(&p); }, worldPos_, viewPos_,
+                            clipPos_, screenPos_);
+
+    referenceSpace_.onChange([this]() { referenceSpaceChanged(); });
+    position_.onChange([this]() { invalidateOutputProperties(); });
     if (camera_) {
-        camera_->onChange([this]() { cameraChanged(); });
+        camera_->onChange([this]() { invalidateOutputProperties(); });
     }
 }
 
 PositionProperty* PositionProperty::clone() const { return new PositionProperty(*this); }
 
-const vec3& PositionProperty::get() const { return positionWorldSpace_; }
+vec3 PositionProperty::get(CoordinateSpace space) const {
+    return convertFromWorld(convertToWorld(position_, referenceSpace_), space);
+}
+
+void PositionProperty::set(const vec3& pos, CoordinateSpace space) {
+    NetworkLock lock(this);
+    referenceSpace_.set(space);
+    position_.set(pos);
+}
+
+void PositionProperty::updatePosition(const vec3& pos, CoordinateSpace sourceSpace) {
+    position_.set(convertFromWorld(convertToWorld(pos, sourceSpace), referenceSpace_));
+}
+
+CoordinateSpace PositionProperty::getCoordinateSpace() const { return referenceSpace_; }
+
+vec3 PositionProperty::getWorldSpaceDirection() const {
+    if (camera_ &&
+        (referenceSpace_ == CoordinateSpace::View || referenceSpace_ == CoordinateSpace::Clip)) {
+        return convertToWorld(position_, referenceSpace_) - camera_->getLookTo();
```

Comment:
No, it is the direction from the origin/look to toward the position.

Reply:
Ok, if we want to have this function in the position property we should update the docs to something like:
```
/* Determines the direction from the origin to the position in the current coordinate space and
 * transforms it to world coordinates. If the property has an associated camera, the origin is the 
 * camera lookTo for view or clip space, otherwise (0, 0, 0).
 * In case of world coordinates, this is equivalent to the position itself.
 * @return position - origin in world space.
 */
```

Reply:
if you return a direction it should probably be normalized? 
is worldspace the only needs space? for this?

Reply:
I think that it would be better to have a new `DirectionPoperty` that has this functionality, but I also understand that it could be outside of the scope of this PR.

Reply:
> if you return a direction it should probably be normalized? is worldspace the only needed space for this?

At the moment yes, since all of our lighting calculations take place in world space and therefore light position and direction is to be specified in world coords.

Reply:
Change function signature to `vec3 getDirection(CoordinateSpace space) const;`. The direction corresponds to the normalized vector between the position and the current offset (either (0,0,0), camera look at, or custom).